opm/simulators/wells/WellInterfaceFluidSystem.cpp

```diff
@@ -162,8 +163,26 @@ activeProductionConstraint(const SingleWellState& ws,
     if (controls.hasControl(Well::ProducerCMode::THP) && currentControl != Well::ProducerCMode::THP) {
         const auto& thp = getTHPConstraint(summaryState);
         double current_thp = ws.thp;
-        if (thp > current_thp)
-            return Well::ProducerCMode::THP;
+        if (thp > current_thp) {
+            bool decrease = true;
+            for (int p = 0; p < number_of_phases_; ++p) {
+                // Currently we use the well potentials here computed before the iterations.
+                // We may need to recompute the well potentials to get a more
+                // accurate check here.
+                decrease = decrease && (-ws.surface_rates[p]) <= ws.well_potentials[p];
```

Comment:
Is the `surface_rates` and `well_potentials` are of different signs here?

Anyway, the logic is that if switching to THP control will result in higher production, we do not switch to THP control. 

As a result, to make the logic easier and code easier to read, we should compare the absolute values and check whether the rate increases. And we discussed the variable should be named `increase` and the logic is the same for injectors and producers.

Reply:
> Is the surface_rates and well_potentials are of different signs here?

Yes. From producers surface_rates are negative, while well_potentials are always positive. 


>Anyway, the logic is that if switching to THP control will result in higher production, we do not switch to THP control.
>As a result, to make the logic easier and code easier to read, we should compare the absolute values and check whether the rate increases. And we discussed the variable should be named increase and the logic is the same for injectors and producers.

I have though a bit more about it, and I prefer the way it is now. 1) I don't think we should check absolute values here. The direction of the flow matters. 2) I don't think reversing the logic and changing the variable name will increase the readability of the code significantly here. I instead added some more comments. I hope that is sufficient.

Reply:
Sure. It is personal preference. My idea is that the code for injectors and producers can basically be almost same. Currently with two different logical way, it takes more time to understand the two sets of code and can be prune to mistakes. 

And here, you do not want some near-zero noisy rates play important roles in determining the simulation behavoir, maybe we should discard those comparisons, either  by comparing the rates of the most important phase we are looking at, or the weighted sum of all rates, or only discarding small values. I do not have an idea which is the best.

Reply:
>  I do not have an idea which is the best.

I don't either. And I don't think we will know before we find a case where the issue you point to is exposed.

Reply:
> I don't either. And I don't think we will know before we find a case where the issue you point to is exposed.

Yeah. That might cause confusion during debugging.  As a first step, maybe we can discard the very small rates first, or using the weighted sum (basically discard the effects of the very small rates), or only check the targeting phase?

The content in this PR should only be checking between the rate controls and THP control following the document of WVFPEXP, so there should be a targeting phase.

Sorry for being so detailed, but it is good the code can be more finely written based on the existing knowledge.  If you disagree, it is okay to agree to disagree.

Reply:
My aim with this PR was not to implement item 4 = yes in WVFPEXP but rather fix an issue that happens when a group controlled well is controlled by group rate equal to zero. The case does not have WVFPEXP and OPM Flow does not support that feature yet. 

`    GR1     ORAT      0.0    1*	    1*    1*    RATE	  YES	  1*	'   '	1*	1*	1*	1*	1*	/
`

and C-1H is under GR1. 

![image](https://user-images.githubusercontent.com/3223024/159264085-d121d084-6262-4f03-8a6e-44b768b7f045.png)
 

With master control will switch between THP and GRUP since Rate THP(rate = 0) violates the thp control of 30. 

`Well C-1H control mode changed from THP->GRUP->THP->GRUP->THP`

![image](https://user-images.githubusercontent.com/3223024/159264668-e7354aeb-c3dc-4ad9-b401-8c6b66b78547.png)

Reply:
> My aim with this PR was not to implement item 4 = yes in WVFPEXP but rather fix an issue that happens 

Yeah. As mentioned above, the logic for the injectors and producers are opposite in this PR, we really hesitate to say yes to that.  And also, for injectors, if we introduce the opposite logic with WVFPEXP, then we will face hug problem when we want to have support for WVFPEXP. And as we discussed, this keyword will take effects even without using it in the commercial simulator.  

> THP and GRUP since Rate THP(rate = 0) 

Then it comes to the question whether we should check THP values with `rate==0`, whether we should do extrapolation for `rate==0`, and also the rate is already zero, maybe we should not switch to any other controls because `rate==0` is the most restrictive control already if the well still operable. If the well is not operable, we should just shut it.  

Whether VFP table is still valid when `rate==0`, I am not expert, but I think it is not designed for those region. Maybe @atgeirr @tskille have some comments regarding that. 

But your fix for the producers part can temporarily fix the problem. My argument is that whether we should introduce an opposite logic for injectors as mentioned in the comments.  


And for the examples your posted figures with, what is the THP values from the reference results from the commercial simulator?